File: source/61.旋转链表.cpp
/*
 * @lc app=leetcode.cn id=61 lang=cpp
 *
 * [61] 旋转链表
 */
#include "cpp_includes.h"
// @lc code=start
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* rotateRight(ListNode* head, int k) {
        if (!head || !head->next) return head;

        auto tail = head;
        int len = 1;
        while (tail->next) {
            tail = tail->next;
            len++;
        }

        k = k % len;
        if (!k) return head;

        auto nhead = head;
        auto pre = head;

        for (int i = 0; i < len-k; i++) {
            pre = nhead;
            nhead = nhead->next;
        }

        pre->next = nullptr;
        tail->next = head;
        return nhead;
    }
};
// @lc code=end

File: source/61.旋转链表.py
#
# @lc app=leetcode.cn id=61 lang=python
#
# [61] 旋转链表
#

# @lc code=start
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution(object):
    def rotateRight(self, head, k):
        """
        :type head: ListNode
        :type k: int
        :rtype: ListNode
        """
        if not head or not head.next: return head
        len_ = 0
        cur = head
        while cur:
            len_ += 1
            cur = cur.next
        
        k %= len_
        if k == 0: return head

        fast, slow = head, head
        while k:
            fast = fast.next
            k -= 1
        
        while fast.next:
            fast = fast.next
            slow = slow.next
        
        newHead = slow.next
        slow.next = None

        fast.next = head
        return newHead
# @lc code=end

